### PR TITLE
Update locals in callbacks

### LINKED ISSUE #760: [Bug] Callbacks can't access local variables
**Description**
From the discussion here #756, callbacks in DQN should have access to local variables, e.g. `done` through `BaseCallback.locals` / `self.locals`. However, this isn't the case. 

**Code example**
```python
from stable_baselines.common.callbacks import BaseCallback
import gym
import stable_baselines as sb
env = gym.make("MountainCar-v0")
class TestCallback(BaseCallback):
    def __init__(self, verbose=0):
        super().__init__(verbose)
    def _on_step(self):
        assert ('done' in self.locals)

agent = sb.DQN('MlpPolicy', env, verbose=1)
agent.learn(100, callback=TestCallback())
```

```bash
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stelios/anaconda3/envs/thesis37/lib/python3.7/site-packages/stable_baselines/deepq/dqn.py", line 224, in learn
    if callback.on_step() is False:
  File "/Users/stelios/anaconda3/envs/thesis37/lib/python3.7/site-packages/stable_baselines/common/callbacks.py", line 89, in on_step
    return self._on_step()
  File "<stdin>", line 5, in _on_step
AssertionError
```

**Workaround**

In `DQN.learn`, line 219, the following results in assertion error.
```python
new_obs, rew, done, info = self.env.step(env_action)

self.num_timesteps += 1
assert 'done' in callback.locals  # fails
# Stop training if return value is False
if callback.on_step() is False:
    break
```
If we call `locals()`, then the assertion succeeds.

```python
new_obs, rew, done, info = self.env.step(env_action)

self.num_timesteps += 1
locals()
assert 'done' in callback.locals  # succeeds
# Stop training if return value is False
if callback.on_step() is False:
    break
```

### PR DESCRIPTION
**Pull request by PartiallyTyped**
- Introduced an `update_locals(locals)` function in `BaseCallback`.
- Added `update_locals(locals())` function call before calling `callback.step`.

## Motivation and Context
- [x] I have raised an issue to propose this change ([required](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) for new features and bug fixes)

Closes #760 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the [changelog](https://github.com/hill-a/stable-baselines/blob/master/docs/misc/changelog.rst) accordingly (**required**).
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have ensured `pytest` and `pytype` both pass (by running  `make pytest` and `make type`).

How/ Where should I list the available parameters in the documentation?

### CODE REVIEW

**Comment by araffin** [comment]
Thanks for the PR, please don't forget to do all the required actions.

Side note: please use double quotes for docstrings and a space after ":". (I'm wondering why codacy did not catch that...)

**Comment by Solliet** [comment]
Since this also requires documentation change, I wanted to ask where to list the available values for each algorithm's callback.

**Comment by araffin** [comment]
> I wanted to ask where to list the available values for each algorithm's callback.

I would say in the callback page, so you have a direct overview. Otherwise, one small note for each algorithm would also work. Do what you think is best ;)

**Comment by Solliet** [comment]
@araffin 
So in the case of ACER/A2C etc, the on_step is called on the runner, not the agent itself. How should this be treated? Should the locals be updated using the values in the runner?

**Comment by araffin** [comment]
> @araffin
> So in the case of ACER/A2C etc, the on_step is called on the runner, not the agent itself. How should this be treated? Should the locals be updated using the values in the runner?

yes, that's an issue, I would update twice, so you can grab all variable using `on_step()` or `on_rollout_end()`
What do you think?

**Comment by Solliet** [comment]
Sounds good, I will put the call right before `callback.on_rollout_end()`.

File: docs/modules/dqn.rst [hunk 1]
```diff
@@ -170,3 +170,84 @@ You can easily define a custom architecture for the policy network:
...
   model.learn(total_timesteps=100000)
+
+
+Callbacks - Accessible Variables 
+-------------
+
+Depending on initialization parameters and timestep, different variables are accessible.
+    +-----------------------------+-----------------------------------------------------+
+    |Variable                     |                                         Availability|
+    +=============================+=====================================================+
```
Comment (araffin) [inline docs/modules/dqn.rst:181]: I would rather group all variables to avoid having too many rows

**Comment by PartiallyTyped** [comment]
Hey everyone, sorry for the delay on this! I was finishing up with school so I didn't have a lot of time to work on this. I will get it done soon.

**Comment by araffin** [comment]
@PartiallyTyped is it ready to review?

File: stable_baselines/common/callbacks.py [hunk 2]
```diff
@@ -50,6 +50,16 @@ def init_callback(self, model: 'BaseRLModel') -> None:
         self.logger = logger.Logger.CURRENT
         self._init_callback()
 
+    def update_locals(self, locals_:Dict[str, Any]):
+        '''
```
Comment (araffin) [inline stable_baselines/common/callbacks.py:54]: Please use double quotes and do not forget `:param locals_` in the docstring
The return type is also missing...

**Comment by PartiallyTyped** [comment]
I have written the docs for DDPG and DQN and the code is ready. Review those to check that the doc okay and I will upload the remaining.

**Comment by PartiallyTyped** [comment]
@araffin I have pushed everything.

**Comment by araffin** [comment]
Ok, I will try to take a look this week. Ping me if you've got no news next monday ;)

**Comment by PartiallyTyped** [comment]
@araffin ping you :)

**Comment by araffin** [comment]
@PartiallyTyped thanks, I put that on my week todo-list ;)

**Comment by PartiallyTyped** [comment]
Umm github committed the change multiple times... Interesting ^^'

**Comment by araffin** [comment]
> Umm github committed the change multiple times... Interesting ^^'


yes, it was quite buggy today: https://www.githubstatus.com/
I had an issue with merging DQN branch too.

**Pull request merged**